re-factoring for 1.3: /lib, /models, /concerns, /services, Rails 3 syntax, other code cleanup

isratrade opened the pull request:
- All tests pass after each commit
- Move files are in separate commits from edit files
- Best to view without whitespace changes https://github.com/theforeman/foreman/pull/765/files?w=1
- Best to review one commit at a time
- Net & Proxy features remain in /lib per @ohadlevy

Commit:
fixes #2411 - move files to /app/services

Commit:
fixes #2411 - move to controller/concerns

Commit:
fixes #2411 - move files in /models to /concerns, /services, /mailers, /observers

Commit:
fixes #2411 - add autoload_paths in application.rb in preparation for file location changes

Commit:
fixes #2411 - move STI models in /models and /lib/foreman/models to sub-folders

Commit:
fixes #2411 - move lib/audit_extensions lib/fog_extensions lib/foreman to models/concerns

Commit:
fixes #2739 - add ActiveSupport::Concern syntax to mixins

Commit:
fixes #2741 - remove unnecessary require statements

Commit:
fixes #2740 - copy code from lib/foreman.rb to initializers/foreman.rb

Commit:
fixes #2741 - rails 3 syntax

Commit:
fixes #2741 - Minitest refactoring

isratrade commented on test/test_helper.rb:
```diff
@@ -33,11 +33,14 @@ class ActiveSupport::TestCase
+    alias_method :assert_not_equal, :refute_equal
+    alias_method :assert_raise,     :assert_raises
+    class <<self
+      alias_method :test,  :it
```
This alias_method makes a normal test name output rather than 0001_anonymous, 0002_anonymous.

isratrade commented on app/views/common/_puppetclasses_or_envs_changed.html.erb:
```diff
@@ -32,19 +32,20 @@
+        <% end %>
+      <% end %>
+    <% end %>
+  </tr>
```
missing < /tr  >was causing warnings output in rake test

isratrade commented on Gemfile:
```diff
@@ -1,5 +1,4 @@
 require File.expand_path('../config/settings', __FILE__)
-require File.expand_path('../lib/regexp_extensions', __FILE__)
```
Why is config/settings in Gemfile and not in an initializer?  When I tried to move it, I got errors?

domcleal replied:
What errors?  I don't get any here when I remove it.

domcleal replied:
Ah, I see.  `bundler.d/jsonp.rb` uses it.

isratrade commented on app/controllers/audits_controller.rb:
```diff
@@ -1,5 +1,3 @@
-require 'audit_extensions'
```
this is already in an initializer, so it can be deleted

isratrade commented:
re-commited.  now tests pass on 1.8.7

domcleal commented on lib/foreman/sti.rb:
```diff
@@ -1,31 +0,0 @@
-    def self.included(base)
-      base.class_eval do
-        class << self
-          # ensures that the correct STI object is created when :type is passed.
```
This potentially useful comment got lost?

isratrade replied:
@domcleal, comment was put back in the files new location.  app/models/concerns/foreman/sti.rb

domcleal commented on app/models/concerns/audit_extensions.rb:
```diff
@@ -0,0 +1,53 @@
+# The audit class is part of audited plugin
+module AuditExtentions
```
Not a bug you introduced, but this should be Exten_s_ions

isratrade commented:
@domcleal, ohad requested that the net and proxy stuff in /lib stay there as it may be extracted into a gem in the future

isratrade commented:
@witlessbird, yes I also agree with validations for the same attribute on the same line.  I meant to do this as a 2nd pass over the code

isratrade commented:
re-committed with validations for the same attribute on the same line new commit removing LOWER()

isratrade commented:
@domcleal, @witlessbird, can you review this branch again at your convenience.

domcleal commented:
I've lost all of the comments I made, because I think they were made on the original commit SHAs, so force-pushing has made them disappear :(  If anybody happens to have the e-mails still, I'd appreciate copies.

Two comments I happen to remember are:
a) lots of things aren't namespaced under Foreman, should we do that too?
b) app/views/unattended/ templates shouldn't be changed because these aren't rendered by Rails, they're plain ERB templates (safemode)

isratrade commented:
@domcleal, I have 4 emails with your comments that are viewable if you click the links above "Show outdated diff"
I also removed two small commits from the original PR that moved some file locations that I moved back to the original location.

domcleal commented:
Hm, I think these were different, that's a pain.  In addition to the two points above, I don't see that the 2801 commit should be part of this PR.

isratrade commented:
@domcleal, I don't understand why to namespace modules or libs under the namespace "Foreman", when the main app is "Foreman".  I think that engines/gems should be namspaced but there is no need for the main app.  What are your thoughts?

isratrade commented:
@domcleal, removed fixes #2801 and reverted changes to app/views/unattended/ templates

isratrade commented:
@domcleal, I plan to change in the future /models/compute_resources/foreman/model/_.rb to just /models/compute_resources/_.rb but I prefer to leave it as is for the initial refactor.

domcleal commented on db/migrate/20090730152224_create_ptables.rb:
```diff
@@ -9,11 +9,11 @@ def self.up
     end
     Ptable.create :name => "RedHat default", :layout =>"zerombr\nclearpart --all --initlabel\nautopart\n"
     Ptable.create :name => "Ubuntu default", :layout => <<EOF
-<% if @host.params['install-disk'] -%>
```
this also shouldn't be changed, as it's a type of provisioning template

domcleal commented:
Regarding namespacing it's app/services/ that mostly bothers me, because we have some stuff under foreman/ and namespaced, but others like SSO that aren't and could conflict with Rails or similar later.  Let's leave it for now, but just something I wanted to raise.

dmitri-d commented on app/controllers/api/v1/reports_controller.rb:
```diff
@@ -33,8 +33,8 @@ def destroy
-        @report = Report.find(max_id, :include => { :logs => [:message, :source] })
+        max_id = Report.my_reports.where(conditions).maximum(:id)
+        @report = Report.includes(:logs => [:message, :source]).find(max_id)
         render :show
```
Mostly a nitpick: I think using report with the latest date would be cleaner and easier to understand.

ohadlevy replied:
afair, on a large table such as reports, using max was much faster

isratrade replied:
@ohadlevy, #max is on an enumerable object.  A quick test in console showed that 

``` ruby
Report.maximum(:id)
```

is much faster than

``` ruby
Report.pluck(:id).max
```

isratrade commented:
@domcleal, I understand now that you want to namespace to avoid conflicts with Rails itself.  Shouldn't having the module named SSO be enough as this is the namespace?

dmitri-d commented on app/models/host/managed.rb:
```diff
@@ -180,11 +173,11 @@ def is_owned_by=(selection)
 
   def clearFacts
-    FactValue.delete_all("host_id = #{id}")
+    FactValue.destroy_all("host_id = #{id}")
```
Good call(s)!

isratrade replied:
yes, then the callbacks such as :dependant => :destroy will be called.

ohadlevy replied:
I'm honestly against any code changing behaviour in this commit.
delete_all is much faster, not sure if we have any callbacks on facts at the moment to use this.

isratrade commented:
re-commited with change in app/controllers/api/v1/reports_controller.rb from :id to :created_at and reverted migration of ptables to include -%>

dmitri-d commented on config/initializers/foreman.rb:
```diff
@@ -1,4 +1,16 @@
+    UUIDTools::UUID.random_create.to_s
+  end
+end
 
```
Should Foreman module definition and Foreman.uuid specifically be moved out from the initializer?

isratrade replied:
@witlessbird, it was originally not in the initializer.  I will move it back out.

isratrade commented:
re-based and re-commited with lib/foreman.rb back

dmitri-d commented:
@domcleal, @isratrade: lgtm

domcleal commented on app/controllers/api/v1/reports_controller.rb:
```diff
@@ -33,8 +33,8 @@ def destroy
         conditions = { :host_id => Host.find_by_name(params[:host_id]).try(:id) } unless params[:host_id].blank?
-        max_id = Report.my_reports.maximum(:id, :conditions => conditions)
-        @report = Report.find(max_id, :include => { :logs => [:message, :source] })
+        max_id = Report.my_reports.where(conditions).maximum(:created_at)
```
This is breaking the functional test for this controller, as maximum(:created_at) is returning the maximum timestamp, not maximum ID.

domcleal commented:
I'm waiting for tests to pass before merging this now, I think we're close enough.

ohadlevy commented on app/models/config_template.rb:
```diff
@@ -4,9 +4,8 @@ class ConfigTemplate < ActiveRecord::Base
-  validates_presence_of :name, :template
-  validates_presence_of :template_kind_id, :unless => Proc.new {|t| t.snippet }
-  validates_uniqueness_of :name
+  validates :name, :template, :presence => true, :uniqueness => true
```
this changes the condition.. why should we ensure template content is not the same?

isratrade replied:
you're right. good catch.  I will change.

ohadlevy commented on app/models/host/managed.rb:
```diff
@@ -180,11 +173,11 @@ def is_owned_by=(selection)
   def clearReports
     # Remove any reports that may be held against this host
-    Report.delete_all("host_id = #{id}")
+    Report.destroy_all("host_id = #{id}")
```
AFAIR, we did this on purpose to make reports deletion faster, later on when we expire reports, we handle the dangling messages.
note that messages are usually removed, so ideally this was suppose to make things faster.

isratrade replied:
ok, I will revert all back to .delete_all.  Only Report has :dependant => :destroy, but if the dangling messages are handled, then it's fine.  however, if the foreign-key branch is merged in the future, this change needs to be re-considered.

ohadlevy commented on app/models/host/managed.rb:
```diff
@@ -195,7 +188,7 @@ def set_token
   def expire_tokens
     # this clean up other hosts as well, but reduce the need for another task to cleanup tokens.
-    Token.delete_all(["expires < ? or host_id = ?", Time.now.utc.to_s(:db), id])
+    Token.destroy_all(["expires < ? or host_id = ?", Time.now.utc.to_s(:db), id])
```
same here, this was done on purpose.

isratrade commented:
re-committed with fixes based on comments and removed commit that changes -%> to %>.  This change can be a separate PR #794

domcleal commented:
I've merged up to and including 4b971988, so minus the Rails 3 syntax change.  Could you please send this as a separate PR so it can be reviewed again?  My aim was to get the file move changes in as we've checked those and then we can do the Rails 3 bit separately as there have been a few comments.

Thanks @isratrade for the work, but separate code moves and code changes into separate PRs in future to ease reviewing.

I think the three PRs we need next are:

a) #794 for the <% changes
b) a PR for [#2801](http://projects.theforeman.org/issues/2801)
c) a PR for the remainder of the Rails 3 changes, [#2741](http://projects.theforeman.org/issues/2741)

isratrade commented:
@domcleal, I agree with your decision to merge the move file location commits and ActiveSupport::Concern commits.  Yes, it in the future I will separate PR's with code moves and code changes.
